### PR TITLE
3.0 Software Development Plan

### PR DESCRIPTION
**Pull request by Perksey**
This creates a roadmap and basic software development plan for 3.0, as well as a support & breaking change policy after the initial 3.0.

### CODE REVIEW

**Comment by Perksey** [comment]
I think this is ready for review (but I may have missed something!) I think we should save roadmap discussion for the Working Group meeting as I think it's better to discuss that in a voice format, but if you have any initial objections to anything else before the meeting let me know.

File: documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+# Silk.NET 3.0
+
+## Goals of 3.0
+
+The key tenets of 3.0 are **portability**, **maintainability**, **usability**, and **performance**. To this end, the following objectives have been identified:
+- Use .NET 6 - the first version of modern .NET to run on the majority of our desired target platforms
+    - (tenet: portability) 
+- Allow Silk.NET's rich abstractions to be integrated into other frameworks rather than being completely standalone.
+    - In order of priority: WPF, WinForms, MAUI, Avalonia (based on demand). (tenet: usability)
```
Comment (HurricanKai) [inline documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md:22]: Weird to list them like this. I'd remove this.

File: documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+- Enhance our input library to add more functionality and future-proof it against future functionality additions.
+    - For more information, see [the Enhanced Input Events proposal](Proposal - Enhanced Input Events.md). (tenet: maintainability)
+- Refactor our input library to work in multiple scenarios and environments
+    - For more information, see [the Multi-Backend Input proposal](Proposal - Multi-Backend Input.md). (tenet: usability)
+
+Silk.NET 3.0 presents us with an opportunity to rethink the entire library taking into account everything we've learnt over the past 2 years of the project's development.
+
+## Development Roadmap
+
+Note that this development roadmap does not take into account unit tests, only functional tests such as experiments. The team should of course stride to add as many tests as possible where possible.
```
Comment (HurricanKai) [inline documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md:38]: Why. One of the important things for 3.0 imo is that it cements it as a production ready product. Unit tests are vital.

File: documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+### 3.0 Preview 1
+
+Before we can do anything, we need to get our brand new generators up and running. In this version:
+- The Scraper works as a minimum viable product. It has minimal support for adding extra attributes for invoking overloaders.
+    - The Khronos APIs in particular will likely be incomplete compared to 2.X in this version. 
+- The Emitter works completely as intended.
+- The Overloader works as a minimum viable product. It doesn't necessarily implement all overloads specified yet.
+- Windowing and Input are implemented for desktop platforms, and have received initial testing.
```
Comment (HurricanKai) [inline documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md:47]: But then they are implemented without using any overloads... That seems weird.

File: documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+- The Scraper is complete.
+- The Overloader is complete.
+- iOS support has been added for Windowing and Input, and have received initial testing on this platform.
+- If time permits, a start has been made on the SIMD APIs in Maths. No work has been done on integrating it into the other Maths types.
+
+3.0 Preview 3 is not a production-ready preview and is very experimental. This preview should aim to release in early October.
+
+### 3.0 Preview 4
+
+This is the first "production-ready" preview and we want users to start integrating into their workloads, so we need to make sure good progress has been made to all goals for the 3.0 update and as many forseeable breaking changes as possible done. In this version:
```
Comment (HurricanKai) [inline documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md:77]: If it's production ready it's a release candidate.

File: documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+- Bugfixes from 3.0 Preview 4
+- A windowing integration for MAUI has been developed and has at least basic OpenGL(ES) support in a state that is as high-performance and as smoothed-out as possible.
+- If time permits, a windowing integration for Avalonia has been developed and has at least basic OpenGL(ES) support. If there is not enough time, this can be pushed to 3.X.
+- SIMD APIs should be integrated into Maths in as many common cases as possible. Ongoing performance improvements may be done in 3.X. This preview should aim to release in early November.
+
+## Problems identified in past development
+
+- We have severely lacking documentation
+    - The intention is that all developers of large amounts of code write implementation documentation and/or "orientation guides" for their codebases informing readers of all major things there is to know in their code.
+    - We should also write documentation containing examples on using as many features of the surface APIs as possible \[for high level utilities\]
```
Comment (HurricanKai) [inline documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md:97]: Can we do this inline?

File: documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    - We will enforce XML documentation in all manually-written utilities and as much as possible in bindings.
+    - If time permits, we should productionize our website powered by Raisin.
+- There's not a lot of planning
+    - We have solved this in the form of the proposal you are reading and all linked proposals: getting all the design done now and documented now, to prevent design debates later down the line. This should reduce friction when actually working on the library.
+    - We have been keeping the working group and key stakeholders in the loop with the 3.0 kickoff (again, see this proposal you are reading)
+    - The team are trying to communicate with eachother and figure out how to distribute work among themselves depending on individual circumstances and free time
+        - Codeowners have been established
```
Comment (HurricanKai) [inline documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md:104]: Have they?

File: documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+- Barrier to entry for external contributors is very high
+    - Documentation should help with this.
+    - We should at least consider introducing something like stylecop to ensure code is readable and easy to navigate.
```
Comment (HurricanKai) [inline documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md:107]: .NET 5 and up ship with enhanced warnings. Also nullability analysis everywhere please.

File: documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+
+If a bug is determined (agreed upon by the majority of maintainers) to be causing massive disruption to the point where the library is borderline unusable in some or all use cases of the library with a considerable proportion of the userbase affected, an out-of-cycle "emergency patch" may be issued on any other Friday between updates.
+
+### Versioning
+
+Any post-3.0, pre-4.0 release will be versioned as follows:
+- The major version will always be 3
+- The minor version will be the number of the monthly update cycle i.e. the first monthly update will be versioned 3.1, the second 3.2, etc...
+- The patch version will always be 0, unless it is an emergency patch in which case it'll be the number of the emergency patch i.e. if an emergency patch is required after the first monthly update the version will be 3.1.1, if another one is required (heaven forbid) in this same cycle it'll be 3.1.2 etc...
+- The revision version will always be 0.
```
Comment (HurricanKai) [inline documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md:131]: Do we want to include the commit hash in future assemblies?

File: documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+
+Users are expected to keep all of the versions of all Silk.NET packages they are using in-sync.
```
Comment (HurricanKai) [inline documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md:133]: Analyzer?

Comment (Perksey) [reply documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md:133]: Perhaps.

```suggestion
Users are expected to keep all of the versions of all Silk.NET packages they are using in-sync. We could write a Roslyn analyser or MSBuild target to help push users to this.
```

File: documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md (hunk 1 above)
Comment (HurricanKai) [reply documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md:22]: True but like order of priority. Idk about that, I'd argue the priorities are different.

Comment (Perksey) [reply documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md:22]: True, I suppose it's subjective. Will remove the in order of priority and just know in my heart that is the order of priority :P 

```suggestion
    - WPF, WinForms, MAUI, Avalonia (tenet: usability) 
```

File: documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md (hunk 2 above)
Comment (HurricanKai) [reply documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md:38]: I don't expect us to do fully functional testing, but I've found a very nice looking lib for approval tests. We can very likely get some actual unit testing working with not as much effort. Imo if we can't test it it's not good. Virtually all of our library is at the heart of a user's library, if we have a bug we have a bug at the heart of the users app.

File: documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md (hunk 3 above)
Comment (HurricanKai) [reply documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md:47]: Hmmm, fair enough.

File: documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md (hunk 4 above)
Comment (HurricanKai) [reply documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md:77]: Right, right, I think I've asked this same thing last time... I keep forgetting 😅

File: documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md (hunk 2 above)
Comment (Perksey) [reply documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md:37]: Oh yeah for sure, I do really want unit testing but I'm also being realistic about our time constraints - we are starting 3.0 development incredibly late!

File: documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md (hunk 5 above)
Comment (HurricanKai) [reply documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md:97]: Mit relevant to this proposal, but maybe you can explain how this works - I trust you are right, but I don't know how we'll do it.

File: documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md (hunk 8 above)
Comment (HurricanKai) [reply documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md:134]: Fair enough.

File: documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md (hunk 5 above)
Comment (Perksey) [reply documentation/proposals/Proposal - 3.0 & 3.X Software Development Plan.md:96]: Sure, documented more.

**Comment by Perksey** [comment]
Can I get a follow-up on this and resolve discussions if you think they're fine?

**Review by HurricanKai — APPROVED** [review]
Other then the code ownership question this is fine.

**Comment by Perksey** [comment]
@ThomasMiz are you happy with this? Please approve on GitHub if you are.

**Review by ThomasMiz — APPROVED** [review]

**Pull request merged**